source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+that frame advances by the length of the pulse.
+
+For frames, everything behaves analogous to qubits in the
+`Delays <delays.html>`_ section of this specification. There are however some
+small differences.
+
+The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
+instruction may also take a list of frames instead of a list of qubits and aligns the time
+of the clocks given as arguments.
+
+``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
+meaning that those clocks are guaranteed to be aligned at the start of the block.
+These blocks also need to have a well-defined length, similar to the ``boxas`` block.
+
+.. code-block:: javascript
+
+   cal {
+     waveform p = ...; // some 100dt waveform
+     frame driveframe1 = newframe(5.0e9, 0);
+     frame driveframe2 = newframe(6.0e9, 0);
+   }
+
+   defcal aligned_gates {
+     // driveframe1 and driveframe2 used in this defcal, so clocks are aligned
+     play(tx0, p, driveframe1);
+     delay[20dt] driveframe1;
+     // Clocks now unaligned by 120dt, so we use a `barrier` to re-align
+     barrier(driveframe1, driveframe2);
+     // `driveframe2` will now play a pulse 120dt after `driveframe1` finishes playing
+     play(tx0, p, driveframe2);
+   }
+
+Examples
+--------
+
+Cross-resonance gate
+~~~~~~~~~~~~~~~~~~~~
+
+
+.. code-block:: javascript
+
+  cal {
+     frame frame0 = newframe(5.0e9, 0);
+  }
+
+  defcal cross_resonance $0 $1 {
+      // Access globally defined channels
+      channel d0 = txch($0, "drive");
+      channel d1 = txch($1, "drive");
+
+      waveform wf1 = gaussian_square(1., 1024dt, 128dt, 32dt);
+      waveform wf2 = gaussian_square(0.1, 1024dt, 128dt, 32dt);
+
+      // phase update some virtual Z gate
+      frame0.phase += pi/2;
+
+      /*** Do pre-rotation ***/
+      {...}
+
+      // generate new frame for second drive that is locally scoped
+      frame temp_frame = copyframe(frame0);
+      temp_frame.phase = frame0.phase + pi/2;
+
+      play(d0, wf1, frame0);
+      play(d1, wf2, temp_frame);
+
+      /*** Do post-rotation ***/
+      {...}
+  }
+
+Geometric gate
+~~~~~~~~~~~~~~
+
+.. code-block:: javascript
+
+  float[32] fq_01 = 5e9; // hardcode or pull from some function
+  float[32] anharm = 300e6; // hardcode or pull from some function
+  cal {
+      frame frame_01 = newframe(fq_01, 0);
+      frame frame_12 = newframe(fq_01 + anharm, 0);
+  }
+
+  defcal geo_gate(angle[32] theta) $0 {
+      // theta: rotation angle (about z-axis) on Bloch sphere
+
+      // Access globally defined channels
+      tx_channel dq = txch($q, “drive”);
+
+      // Assume we have calibrated 0->1 pi pulses and 1->2 pi pulse
+      // envelopes (no sideband)
+      waveform X_01 = {...};
+      waveform X_12 = {...};
+      float[32] a = sin(theta/2);
+      float[32] b = sqrt(1-a**2);
+
+      // Double-tap
+      play(dq, scale(a, X_01), frame_01);
```

Comment:
Does `scale` function exist anymore? Don't see it...

Reply:
Yeah, it's gone but I think it's valuable to re-introduce. Made a comment to that effect in the "Waveform" section.

Reply:
Reintroduced.